Add `show_fps` argument to display FPS on video/stream predictions

PrashantDixit0 opened the pull request:
<!--
Thank you 🙏 for your contribution to [Ultralytics](https://www.ultralytics.com/) 🚀! Your effort in enhancing our repositories is greatly appreciated. To streamline the process and assist us in integrating your Pull Request (PR) effectively, please follow these steps:

1. Check for Existing Contributions: Before submitting, kindly explore existing PRs to ensure your contribution is unique and complementary.
2. Link Related Issues: If your PR addresses an open issue, please link it in your submission. This helps us better understand the context and impact of your contribution.
3. Elaborate Your Changes: Clearly articulate the purpose of your PR. Whether it's a bug fix or a new feature, a detailed description aids in a smoother integration process.
4. Ultralytics Contributor License Agreement (CLA): To uphold the quality and integrity of our project, we require all contributors to sign the CLA. Please confirm your agreement by commenting below:

    I have read the CLA Document and I sign the CLA

For more detailed guidance and best practices on contributing, refer to our ✅ [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Your adherence to these guidelines ensures a faster and more effective review process.
-->

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Add an optional CLI/config flag to overlay real-time FPS on video and stream predictions.

### 📊 Key Changes
- Introduces a new `show_fps` configuration option in `default.yaml` (default `False`) to control FPS display.
- Imports `time` and adds `self.last_frame_process_time` state in `Predictor` for FPS calculation.
- Updates `Predictor.show()` to compute simple frame-to-frame FPS and render it on the output frame when `show_fps` is enabled.

### 🎯 Purpose & Impact
- 🧩 Gives users a lightweight way to monitor inference FPS directly on video/stream outputs.
- 🎥 Improves usability for benchmarking and debugging real-time performance without external tools.
- ⚙️ Keeps the feature opt-in via config/CLI, avoiding overhead or overlays for users who do not need FPS.

PrashantDixit0 committed:
add: cli argument to show FPS over stream/video

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Initialize the FPS timing only after the first rendered frame so that the overlay never shows a misleading 0 FPS reading on startup.

💬 Posted 1 inline comment

PrashantDixit0 committed:
Update ultralytics/engine/predictor.py

Co-authored-by: Ultralytics Assistant <135830346+UltralyticsAssistant@users.noreply.github.com>
Signed-off-by: Prashant Dixit <54981696+PrashantDixit0@users.noreply.github.com>

PrashantDixit0 committed:
fix for not showing 0 FPS

PrashantDixit0 committed:
fix for not showing 0 FPS

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

onuralpszr committed:
Merge branch 'main' into show_fps

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please fix the FPS overlay logic so that each frame computes and displays FPS instead of only the initial frame overwriting `last_frame_process_time`. Once addressed, the new `show_fps` option should work as intended.

💬 Posted 1 inline comment

PrashantDixit0 committed:
fix

PrashantDixit0 committed:
Merge branch 'show_fps' of github.com:PrashantDixit0/ultralytics into show_fps

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 3

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Guard FPS rendering so `Predictor.show()` does not crash when `self.dataset` lacks `.mode`, and skip the first overlay until a baseline timestamp exists to avoid misleading FPS spikes.

💬 Posted 2 inline comments

PrashantDixit0 committed:
fix: suggested by reviewer

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 4

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Guard `show_fps` usage inside `Predictor.show()` so legacy predictors that don’t populate the new argument do not crash when displaying frames.

💬 Posted 1 inline comment

PrashantDixit0 committed:
Update ultralytics/engine/predictor.py

Co-authored-by: Ultralytics Assistant <135830346+UltralyticsAssistant@users.noreply.github.com>
Signed-off-by: Prashant Dixit <54981696+PrashantDixit0@users.noreply.github.com>

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 5

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Please harden the FPS overlay by switching to a monotonic timer and guarding the divisor so the visualization can’t crash when consecutive frames timestamp the same.

💬 Posted 2 inline comments

PrashantDixit0 committed:
fix: ZeroDivisionError handling

Co-authored-by: Ultralytics Assistant <135830346+UltralyticsAssistant@users.noreply.github.com>
Signed-off-by: Prashant Dixit <54981696+PrashantDixit0@users.noreply.github.com>

PrashantDixit0 committed:
fix: time calculation

Co-authored-by: Ultralytics Assistant <135830346+UltralyticsAssistant@users.noreply.github.com>
Signed-off-by: Prashant Dixit <54981696+PrashantDixit0@users.noreply.github.com>

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 6

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Indent `show_fps` under the visualization block so the new flag is parsed correctly.

💬 Posted 1 inline comment

PrashantDixit0 commented:
@glenn-jocher Please review it and let me know if any change is required to merge this PR.

UltralyticsAssistant commented on ultralytics/engine/predictor.py:
```diff
@@ -486,8 +488,22 @@ def save_predicted_images(self, save_path: Path, frame: int = 0):
         im = self.plotted_img
+
+        # FPS calculation
+        if getattr(self.args, "show_fps", False):
```
💡 **MEDIUM**: `self.last_frame_process_time` is shared across every stream/video handled by a single `Predictor`. When multiple sources are processed in the same loop, the FPS overlay on stream B uses the elapsed time since stream A’s frame, causing wildly incorrect numbers (<1 FPS even when running in real time). Please store timestamps per window/source (e.g., a dict keyed by `p`) so each overlay reflects the correct frame cadence for that stream.

UltralyticsAssistant commented on ultralytics/engine/predictor.py:
```diff
@@ -486,8 +488,22 @@ def save_predicted_images(self, save_path: Path, frame: int = 0):
+                        im, f"FPS: {fps:.2f}", (20, 40), cv2.FONT_HERSHEY_SIMPLEX, 1.0, (0, 255, 0), 2, cv2.LINE_AA
+                    )
+
+                self.last_frame_process_time = current_frame_process_time
```
📝 **LOW**: `self.last_frame_process_time` is never reset when `show_fps` is toggled off or when the dataset mode switches away from video/stream. Re‑enabling the flag after a pause will display FPS computed over the idle period, often near zero. Resetting the timestamp to `0.0` whenever FPS drawing is skipped will avoid misleading first-frame values.

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 7

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Nice addition—config stays opt‑in and the overlay logic is lightweight. Please make the FPS state per stream (not global) and reset the timestamp when FPS isn’t drawn so the overlay stays accurate when sources or settings change.

💬 Posted 2 inline comments

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
fix: fps leak for new source

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

PrashantDixit0 committed:
docs: added visualization docs for show_fps

PrashantDixit0 committed:
Merge branch 'show_fps' of github.com:PrashantDixit0/ultralytics into show_fps

PrashantDixit0 committed:
docs: added visualization docs for show_fps

PrashantDixit0 commented:
I have made suggested changes. Please review it

glenn-jocher commented:
Thanks for the quick updates, @PrashantDixit0—please double‑check that `yolo predict` with `show_fps=True` runs cleanly on a short video and that the test suite passes locally, and we’ll queue this for the next review pass.

PrashantDixit0 commented:
I have tested it `yolo predict` with `show_fps=True` param and it works as expected, showing FPS in the left corner of the stream.

PrashantDixit0 committed:
Merge branch 'main' into show_fps

onuralpszr committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

onuralpszr committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

Signed-off-by: Prashant Dixit <54981696+PrashantDixit0@users.noreply.github.com>

PrashantDixit0 committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

PrashantDixit0 committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 commented:
@glenn-jocher Any update on merging this ?

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

UltralyticsAssistant commented on ultralytics/engine/predictor.py:
```diff
@@ -502,6 +506,19 @@ def save_predicted_images(self, save_path: Path, frame: int = 0):
+            mode = getattr(self.dataset, "mode", None)
+            if mode in {"video", "stream"}:
+                current_frame_process_time = time.perf_counter()
+                if self.last_frame_process_time != 0:
```
⚠️ **HIGH**: `self.last_frame_process_time` is shared across every call to `show()`, so the FPS overlay becomes incorrect for common multi-stream cases and when moving between different video files/windows. In `stream` mode, each source updates the same timestamp, so later frames measure the gap from a different stream instead of their own previous frame. This needs to be tracked per source/window key (for example `p` or source index) rather than as a single scalar.

UltralyticsAssistant reviewed (commented):
## 🔍 PR Review 8

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The change is small and generally fits the existing predictor flow, but the current FPS state is global, which makes the reported value wrong for multi-stream inputs and across separate videos/windows. I didn't spot other clear issues in the visible diff.

💬 Posted 1 inline comment

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

glenn-jocher committed:
Merge remote-tracking branch 'origin/main' into show_fps

# Conflicts:
#	ultralytics/engine/predictor.py

glenn-jocher commented:
Hi @PrashantDixit0, I've merged the latest `main` into your branch to resolve the conflict that had built up in `BasePredictor.show()` (`main` recently changed the window-creation guard from `platform.system() == "Linux"` to `platform.system() in {"Linux", "Windows"}`). I kept both that update and your FPS overlay block, so the branch is current again and the effective diff is still just your intended 18 lines across `default.yaml` and `predictor.py`. `import ultralytics` and an AST parse of the changed file both pass locally.

A few concrete items to finish it off:

1. **Register the flag as a bool.** `show_fps` isn't in `CFG_BOOL_KEYS` in `ultralytics/cfg/__init__.py` (line ~253). It works today because `smart_value("True")` already returns a real bool on the CLI, but adding `show_fps` there alongside `show`, `show_boxes`, `show_labels`, and `show_conf` gives it the same type-validation behavior as the other `show_*` flags.
2. **The docs + test Glenn asked for earlier.** A one-line mention of `show_fps` in the predict-mode docs and a tiny test exercising `yolo predict ... show_fps=True` would close out the last checklist items.

Two behavioral notes (not blockers, just so the scope is intentional): the overlay is drawn inside `show()`, so FPS only appears when `show=True` — a save-only run (`save=True show=False`) won't get it. And because `show()` runs just before `save_predicted_images()` and `cv2.putText` mutates `self.plotted_img` in place, the FPS text is also burned into the saved video/frames when `show`, `save`, and `show_fps` are all enabled. Both are reasonable, just worth confirming they match what you intend.

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
added in doc and test

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

PrashantDixit0 commented:
Hi @glenn-jocher 

I've addressed both points mentioned above in a recent push. I've added `show_fps` so that the real-time FPS can be seen during processing. Let me know if you'd like me to include it in the frames that are written to disk.

glenn-jocher commented:
Thanks @PrashantDixit0. No need to expand this PR to add save-only FPS overlays; let’s keep the scope as real-time display via `show=True show_fps=True`, with the existing saved-output behavior only where `show()` mutates `self.plotted_img` before saving. One small cleanup before final review: please move the new docs note out of the `Results.plot()` parameter table, since `show_fps` is a predict/display config flag rather than a `result.plot()` argument, and mention it under the predict visualization args instead.

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
moved show_fps in visualization args section

PrashantDixit0 commented:
Hi @glenn-jocher ,
Moved `show_fps` from `Results.plot()` parameter table to predict Visualization args

PrashantDixit0 committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

glenn-jocher commented:
Thanks @PrashantDixit0, confirmed the `show_fps` documentation is now under predict visualization arguments and all CI checks are green. Nothing else is needed from your side; this is ready for final maintainer review.

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Merge branch 'main' into show_fps

PrashantDixit0 committed:
Merge branch 'main' into show_fps

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher commented:
Hi @PrashantDixit0, thank you for the contribution! After review we've decided not to add a dedicated `show_fps` argument — FPS is directly computable from the per-image inference speeds Ultralytics already reports (`results[0].speed`, and the live ms readouts during streaming inference), so a separate display flag adds an argument without adding information. Closing as not planned. We appreciate the effort and hope to see more contributions from you!
